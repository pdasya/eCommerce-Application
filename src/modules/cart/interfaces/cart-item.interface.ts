export interface ICartItem {
  id: string;
  sku: string;
  name: string;
  imageUrl: string;
  quantity: number;
  initialPrice: number;
  finalPrice: number;
}
