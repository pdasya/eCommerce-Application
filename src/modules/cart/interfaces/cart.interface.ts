import { ICartItem } from './cart-item.interface';

export interface ICart {
  id: string;
  version: number;
  items: ICartItem[];
  currency: string;
  initialPrice: number;
  discount: number;
  finalPrice: number;
}
