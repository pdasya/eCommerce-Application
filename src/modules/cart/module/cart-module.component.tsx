import { Button, Grid, Typography } from '@mui/material';
import { Box } from '@mui/system';
import React, { FC, useEffect, useState } from 'react';
import { toast } from 'react-toastify';
import { selectCart } from '@store/cart/cart.slice';
import { useAppSelector } from '@hooks/use-app-selector.hook';
import { cartService } from '@modules/cart/services/cart.service';
import { selectAuthorization } from '@store/auth/auth.slice';
import { CartHeader } from '../components/cart-header/cart-header';
import { CartItemList } from '../components/cart-item-list/cart-item-list';
import { PromoCodeForm } from '../components/cart-promocode-form/cart-promocode-form';
import { CartActions } from '../components/cart-actions/cart-actions';
import { ICartItem } from '../interfaces/cart-item.interface';
import * as styles from './cart-module.component.module.scss';

export const CartModule: FC = () => {
  const cart = useAppSelector(selectCart);
  const isAuthorized = useAppSelector(selectAuthorization);

  const [loading, setLoading] = useState(false);
  const [promoCode, setPromoCode] = useState('');
  const [promoCodeId, setPromoCodeId] = useState('');
  const [promoCodeState, setPromoCodeState] = useState(false);
  const [promoError, setPromoError] = useState('');

  const getPromoName = async () => {
    try {
      const results = await cartService.getPromoCodeInfo();
      if (results) {
        if (results.promoName) {
          setPromoCode(results.promoName);
          setPromoCodeId(results.ID);
          setPromoCodeState(true);
        }
      }
    } catch (error) {
      toast.error(error);
    }
  };

  useEffect(() => {
    getPromoName();
  }, [promoCodeState]);

  useEffect(() => {
    setLoading(true);
    cartService
      .synchronize()
      .catch(error => toast.error(error))
      .finally(() => setLoading(false));
  }, [isAuthorized]);

  const addMockItems = async () => {
    setLoading(true);

    const items = [
      {
        sku: 'yamato-sauces-3_1',
        quantity: 1,
      },
      {
        sku: 'itoen-tea-1_1',
        quantity: 1,
      },
      {
        sku: 'itoen-tea-4_1',
        quantity: 1,
      },
      {
        sku: 'itoen-tea-3_1',
        quantity: 1,
      },
    ];

    try {
      await cartService.addCartItem(items);
      setLoading(false);
    } catch (error) {
      toast.error('Error fetching cart:', error);
    } finally {
      setLoading(false);
    }
  };

  const promoCodeRemover = async () => {
    if (promoCode) {
      await cartService.removePromoCode(promoCodeId);
    }
    setPromoCode('');
    setPromoCodeState(false);
  };

  const removeItemHandler = async (item: ICartItem) => {
    try {
      await cartService.removeCartItem(item);
      toast.success(`Item ${item.name} is successfully deleted from the cart`);
    } catch (error) {
      toast.error(`Error deleting item`);
    }
  };

  const incrementItemHandler = async (item: ICartItem) => {
    try {
      cartService.updateCartItemQuantity([{ ...item, quantity: item.quantity + 1 }]);
      toast.success(`Item ${item.name} quantity successfully changed`);
    } catch (error) {
      toast.error(`Error updating quantity: ${error}`);
    }
  };

  const decrementItemHandler = async (item: ICartItem) => {
    try {
      cartService.updateCartItemQuantity([{ ...item, quantity: item.quantity + -1 }]);
      toast.success(`Item ${item.name} quantity successfully changed`);
    } catch (error) {
      toast.error(`Error updating quantity: ${error}`);
    }
  };

  const handleCartClearItems = async () => {
    if (window.confirm('Are you sure you want to clear the cart?')) {
      setLoading(true);
      try {
        await promoCodeRemover();
        await cartService.deleteCart();
        toast.success('Cart has been cleared');
      } catch (error) {
        toast.error('Error clearing cart');
      } finally {
        setLoading(false);
      }
    }
  };

  const handlePromoCodeChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { value } = e.target;
    setPromoCode(value);
    setPromoError('');
  };

  const handlePromoCodeApplyItems = async () => {
    try {
      await cartService.applyPromoCode(promoCode);
      toast.success(`Promo code "${promoCode}" is successfully applied`);
      setPromoCodeState(true);
    } catch (error) {
      setPromoError('Invalid promo code');
      toast.warning(`Unable apply promo code: ${error.message}`);
    }
  };

  const handlePromoCodeDeleteItems = async () => {
    try {
      await promoCodeRemover();
      toast.success(`Promo code "${promoCode}" is successfully deleted`);
    } catch (error) {
      setPromoError('Invalid promo code');
      toast.warning(`Unable deleted promo code: ${error.message}`);
    }
  };

  const handlePromoCodePressEnter = (e: React.KeyboardEvent) => {
    if (e.key === 'Enter' && promoCode.length > 0) {
      handlePromoCodeApplyItems();
    }
  };

  return (
    <Grid container spacing={2} className={styles.cartModuleWrapper}>
      <CartHeader />
      <Grid item xs={12}>
        <Box className={styles.cartModuleContent}>
          {loading ? (
            <Typography variant="body1" gutterBottom>
              Loading...
            </Typography>
          ) : cart.items.length === 0 ? (
            <Typography variant="body1" gutterBottom>
              Your cart is empty
            </Typography>
          ) : (
            <>
              <CartItemList
                cartItems={cart.items}
                onIncrement={incrementItemHandler}
                onDecrement={decrementItemHandler}
                onRemove={removeItemHandler}
              />
              <Box textAlign="right" marginTop={2}>
                {cart.initialPrice !== cart.finalPrice && (
                  <Typography variant="body2" style={{ textDecoration: 'line-through' }}>
                    ${cart.initialPrice.toFixed(2)}
                  </Typography>
                )}
                <Typography variant="h6">Total Price: ${cart.finalPrice.toFixed(2)}</Typography>
              </Box>
              <PromoCodeForm
                promoCode={promoCode}
                promoCodeState={promoCodeState}
                promoError={promoError}
                handlePromoCodeChange={handlePromoCodeChange}
                handlePromoCodePressEnter={handlePromoCodePressEnter}
                handlePromoCodeApply={handlePromoCodeApplyItems}
                handlePromoCodeDeleteItems={handlePromoCodeDeleteItems}
              />
              <CartActions handleCartClear={handleCartClearItems} />
            </>
          )}
          {cart.items.length === 0 && (
            <Button variant="contained" color="info" onClick={addMockItems}>
              Add Cart Items
            </Button>
          )}
        </Box>
      </Grid>
    </Grid>
  );
};
