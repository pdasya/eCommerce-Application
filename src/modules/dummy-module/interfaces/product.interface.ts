export interface IProduct {
  id: number;
  name: string;
  price: number;
  size: 's' | 'm' | 'l';
}
