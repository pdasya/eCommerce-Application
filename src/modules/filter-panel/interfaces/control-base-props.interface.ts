import { ForwardedRef } from 'react';
import { IControlHandle } from './control-handle.interface';

export interface IControlBaseProps<T> {
  name: string;
  ref?: ForwardedRef<IControlHandle>;
  onChange?: (payload: [string, T]) => void;
  onDefaultChange?: (isDefault: boolean) => void;
}
