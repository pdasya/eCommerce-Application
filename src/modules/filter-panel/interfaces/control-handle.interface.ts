export interface IControlHandle {
  reset: () => void;
  getIsDefault: () => boolean;
}
