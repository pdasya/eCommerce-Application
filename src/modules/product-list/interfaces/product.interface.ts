export interface IProduct {
  id: number;
  title: string;
  description: string;
  imageSrc: string;
  price: number;
  currency: string;
  discountPrice?: number;
}
