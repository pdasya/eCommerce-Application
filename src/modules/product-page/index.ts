export { Product } from './product-page-module/product-page.module';
