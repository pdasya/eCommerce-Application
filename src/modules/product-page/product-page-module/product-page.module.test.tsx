import React from 'react';
import { act } from '@testing-library/react';
import '@testing-library/jest-dom';
import { select } from '@store/product/product.slice';
import { renderWithProviders } from '@/utils/render-with-providers.test-util';
import { ISingleProduct } from '@/interfaces/interfaces';
import { Product } from './product-page.module';

describe('Product page module', () => {
  test('snapshot match', () => {
    const mockProduct: ISingleProduct = {
      currency: 'currency',
      currentPrice: 'price',
      description: 'description',
      images: [
        { dimensions: { w: 100, h: 100 }, url: 'img-url-1', label: 'img-label-1' },
        { dimensions: { w: 200, h: 200 }, url: 'img-url-2', label: 'img-label-2' },
      ],
      id: 'id',
      attributes: [
        { name: 'attribute1', value: 'value1' },
        { name: 'attribute2', value: 'value2' },
      ],
      title: 'title',
      discountPrice: 'description',
    };

    const { store, asFragment } = renderWithProviders(<Product />);

    expect(asFragment()).toMatchSnapshot('product-page-module-snapshot-1');

    act(() => store.dispatch(select(mockProduct)));

    expect(asFragment()).toMatchSnapshot('product-page-module-snapshot-2');
  });
});
