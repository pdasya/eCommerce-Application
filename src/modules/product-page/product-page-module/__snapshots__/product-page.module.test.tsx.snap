// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product page module snapshot match: product-page-module-snapshot-1 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 css-1fyyp8j-MuiGrid-root"
    >
      <div
        class="f-carousel"
      >
        <div
          class="f-carousel"
        >
          <a
            data-fancybox="gallery"
            href="./assets/images/no-image.jpg"
          >
            <img
              alt="noImage"
              src="./assets/images/no-image.jpg"
            />
          </a>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 css-1fyyp8j-MuiGrid-root"
    >
      <div
        class="MuiGrid-root css-ndpptv-MuiGrid-root"
      >
        <h4
          class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
        />
        <div />
        <div>
          <p
            class=""
          >
            0
          </p>
          <p>
            0
          </p>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-sghohy-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Add to cart
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ShoppingCartIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2M1 2v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.14 0-.25-.11-.25-.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.58-6.49c.08-.14.12-.31.12-.48 0-.55-.45-1-1-1H5.21l-.94-2zm16 16c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Product page module snapshot match: product-page-module-snapshot-2 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 css-1fyyp8j-MuiGrid-root"
    >
      <div
        class="f-carousel"
      >
        <div
          class="f-carousel"
        >
          <div
            class="f-carousel__slide"
            data-fancybox="gallery"
            data-src="img-url-1"
            data-thumb-src="img-url-1"
          >
            <img
              alt="img-label-1"
              src="img-url-1"
            />
          </div>
          <div
            class="f-carousel__slide"
            data-fancybox="gallery"
            data-src="img-url-2"
            data-thumb-src="img-url-2"
          >
            <img
              alt="img-label-2"
              src="img-url-2"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 css-1fyyp8j-MuiGrid-root"
    >
      <div
        class="MuiGrid-root css-ndpptv-MuiGrid-root"
      >
        <h4
          class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
        >
          title
        </h4>
        <div />
        <div>
          <p
            class=""
          >
            pricecurrency
          </p>
          <p>
            descriptioncurrency
          </p>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-sghohy-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Add to cart
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ShoppingCartIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2M1 2v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.14 0-.25-.11-.25-.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.58-6.49c.08-.14.12-.31.12-.48 0-.55-.45-1-1-1H5.21l-.94-2zm16 16c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <nav
          aria-labelledby="nested-list-subheader"
          class="MuiList-root MuiList-padding css-822hb5-MuiList-root"
        >
          <div
            class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiListItemText-root css-tlelie-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
              >
                description
              </span>
            </div>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ExpandMoreIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </div>
        </nav>
        <nav
          aria-labelledby="nested-list-subheader"
          class="MuiList-root MuiList-padding css-822hb5-MuiList-root"
        >
          <div
            class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiListItemText-root css-tlelie-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
              >
                attribute1
              </span>
            </div>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ExpandMoreIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </div>
        </nav>
        <nav
          aria-labelledby="nested-list-subheader"
          class="MuiList-root MuiList-padding css-822hb5-MuiList-root"
        >
          <div
            class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiListItemText-root css-tlelie-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
              >
                attribute2
              </span>
            </div>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ExpandMoreIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </div>
        </nav>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
